library/pod-security-policy/users/template.yaml

```diff
@@ -102,8 +102,17 @@ spec:
 
         get_user_violation(params, container) = msg {
           not get_field_value("runAsUser", container, input.review)
+          not get_field_value("runAsNonRoot", container, input.review)
```

Comment:
I don't think we need to test this here, as the `params.rule != "runAsNonRoot"` line below means that this check will only run for `MustRunAs`, which does require a specific user be provided.

`RunAsNonRoot` with an undefined user is checked in the extra `get_user_violation` you've added below.

Reply:
Like so?

If `runAsUser` is present, it should always be checked, the `accept_users` func sorts out if it's ok:

```rego
get_user_violation(params, container) = msg {
  rule := params.rule
  provided_user := get_field_value("runAsUser", container, input.review)
  not accept_users(rule, provided_user)
  msg := sprintf("Container %v is attempting to run as disallowed user %v. Allowed runAsUser: %v", [container.name, provided_user, params])
}
```

When we ask for a specific user range via `MustRunAs`, `runAsUser` field should be present:

```rego
get_user_violation(params, container) = msg {
  not get_field_value("runAsUser", container, input.review)
  params.rule = "MustRunAs"
  msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser", [container.name])
}
```

If `MustRunAsNonRoot` and user doesn't add `runAsUser`, or `runAsNonRoot`: fail:

```rego
get_user_violation(params, container) = msg {
  params.rule = "MustRunAsNonRoot"
  not get_field_value("runAsUser", container, input.review)
  not get_field_value("runAsNonRoot", container, input.review)
  msg := sprintf("Container %v is attempting to run without a required securityContext/runAsNonRoot or securityContext/runAsUser != 0", [container.name])
}
```